### DISCUSSION TITLE
Look and Feel for Builder

### DISCUSSION THREAD

**Discussion by rabbibotton** [rabbibotton/clog#228] [Ideas]
I still have a few small areas to fill out for the Builder's v1 branch (templates for individual panels added to projects, improvements to documentation of how DB controls work, improving the properties panel, a bit more work on the editor and REPL windows) for what I consider to be my minimum "feature complete" for version 1 builder (no worries 2.0 design already in the works, so far data modeling, many more custom controls, "views", and more).

I have already been using the builder to develop even non CLOG projects, as at this point it is a great IDE for Common Lisp over all but I'd like to get it to a point that puts it on par with emacs.

So this Is about improving the workflow of builder, better keyboard handling, the actual layout of panels, etc. I would love to see mock ups of how you feel builder's IDE should look, 1. when editing panels. 2 when editing lisp files, etc

Feel free to go nuts with your ideas. I value them and your ideas - How would you want such an IDE to look :)

This is the current L&F as of today (projects, panels, dirs and editor, repl, M-. System Browser):

![Screen Shot 2022-09-04 at 10 28 27 AM](https://user-images.githubusercontent.com/11616789/188318731-b0744578-98bb-452c-8a8f-881dd3804629.png)
![Screen Shot 2022-09-04 at 10 29 26 AM](https://user-images.githubusercontent.com/11616789/188318733-715fb606-cfa3-42e6-892b-a3f2dd0b5b7a.png)
![Screen Shot 2022-09-04 at 10 30 48 AM](https://user-images.githubusercontent.com/11616789/188318736-010e25c6-831a-416d-acdc-167c2399516f.png)
![Screen Shot 2022-09-04 at 10 46 22 AM](https://user-images.githubusercontent.com/11616789/188319446-28e52e40-34ae-48cf-b532-a8c2693aa910.png)
![Screen Shot 2022-09-04 at 10 49 55 AM](https://user-images.githubusercontent.com/11616789/188319635-095d1986-cff1-49f4-b1b0-dfc6d6607e96.png)

**Comment by kaeland** [comment]
Hi Dr. Botton! The design of Clog looks really great! I've recently taken on the endeavor of learning Clog, and I'm hoping to be able to create some really useful applications. Thank you for building Clog!

**Comment by kaeland** [comment]
I think some really cool features would be: 

1. A graphical way to manipulate css [flexbox](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Flexbox) and [css grid](https://developer.mozilla.org/en-US/docs/Web/CSS/grid) controls for any new clog gui menu. The current snap-to-position controls work really well, and I'm wondering if there's any room for flexbox and css-grid to support this too! 
2. Filtering on tools and controls. I'd love to be able to quickly find a specific control (background-color, width, height, etc), or tools (Label, Button, Table, etc.) when using Clog Builder. Maybe having an input box right above each of these windows where a user can search through the available controls would work? 
3. In the Mechanical Design world, tools like Fusion360 allow you to copy components from one project to another. In future versions of the builder, it'd be really cool if a user could group buttons, inputs, etc into useful widgets, and then copy them quickly to be used in another Clog project! So, you'd essentially be able to have a directory of useful components you've made along your journey of being a Clogger! 
4. This one isn't really a feature, but I can't state highly enough how much I enjoy your video tutorials! They really are helpful! Being able to see your workflow when building a new gui tool provides a user with a great reference for getting up to speed quickly. If there's any time within your schedule, I'd love to see more video tutorials! Here are a few suggestions/questions I've come up with while using the Clog Builder. They may help with tutorial content: 
- What is Clog? 
- How is Clog different from modern web developement (ReactJS, VueJS)?
- How to build a form
- How to use database controls (One Row, Table Many Rows, etc.)
- Authenticating a new user 

Over time, I believe other users would be able to create new learning content as more people use Clog for development. I'd also like to support this effort as well! I'm learning my way around the Builder, but I'd love to create tutorials for others as time goes on. :D 

Once again Dr. Botton, thank you for all of your effort in building Clog!

**Reply by rabbibotton** [reply]
1 and 2 - awesome ideas, they will probably go in to builder 2
3 - Already can do that :) - drop a div, then while the div (easiest is to make it static or drag to 0,0) then shift click anything you want to add as children of that div (you can also just start with a pre-existing parent control) - copy it and now you can paste it in to notepad, another panel, another instance of builder etc. I for sure have to do a video tutorial to show the cut and paste coolness of the gui builder.
4 - Agreed and I figured a way to be able to do more - record view first and _then_ record the audio, it is much smoother that way for me and will I have _many_ kids and it is really loud... but at least I can find some quiet here and there to add audio later. I hope to do a bunch of intro practical vids soon.

I look forward to others doing vids and docs, plug-ins, projects etc. I think every one should get a chance to enjoy Common Lisp and CLOG.

**Comment by imadhui** [comment]
I didn't try the CLOG IDE yet. Does it support SLIME or SLY?

**Reply by rabbibotton** [reply]
You can start the CLOG Builder from a SLIME or SLY REPL using (clog-tools:clog-builder) or you can start it up directly in SBCL as it has its own REPL as well that uses swank directly.
